### ISSUE TITLE
Errors thrown while using AsyncSerialExecutionStrategy can cause stuck threads

### ISSUE THREAD

**Issue by rlhagen** [graphql-java/graphql-java#3548]
**Describe the bug**
If Errors are thrown in a DataFetcher when using AsyncSerialExecutionStrategy then it's possible to craft a query that causes the thread to be stuck indefinitely. 

**Possible Fix**
Catch Throwable [here](https://github.com/graphql-java/graphql-java/blob/20.x/src/main/java/graphql/execution/Async.java#L169) seems to give the correct behavior completing the future exceptionally which eventually hands it back to the JVM.

**To Reproduce**
Run the following junit test  and observe that `errorsShouldBeNotBeSwallowed` will never terminate.

```java
import static graphql.schema.idl.RuntimeWiring.newRuntimeWiring;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.Arrays;
import java.util.concurrent.CompletionException;

import org.junit.Test;

import graphql.ExecutionResult;
import graphql.GraphQL;
import graphql.execution.AsyncSerialExecutionStrategy;
import graphql.schema.DataFetcher;
import graphql.schema.GraphQLSchema;
import graphql.schema.idl.RuntimeWiring;
import graphql.schema.idl.SchemaGenerator;
import graphql.schema.idl.SchemaParser;
import graphql.schema.idl.TypeDefinitionRegistry;

public class AsyncSerialExecutionStrategyTest {

	@Test(expected = CompletionException.class)
	public void errorsShouldBeNotBeSwallowed() {
		ExecutionResult executionResult = run(environment -> {
			throw new LinkageError("This error should be caught to fail the future");
		});
	}

	@Test
	public void exceptionsShouldCollected() {
		ExecutionResult executionResult = run(environment -> {
			throw new RuntimeException("This exception should be collected");
		});
		assertEquals(1, executionResult.getErrors().size());
		assertTrue(executionResult.getErrors().get(0).getMessage().contains("This exception should be collected"));
	}

	private ExecutionResult run(DataFetcher<?> fetcher) {
		String schema = "type Query{animals: [Animal] plants:[Plant]} type Animal{name: String} type Plant{name: String}";
		SchemaParser schemaParser = new SchemaParser();
		TypeDefinitionRegistry typeDefinitionRegistry = schemaParser.parse(schema);

		RuntimeWiring runtimeWiring = newRuntimeWiring()
				.type("Query", builder -> builder.dataFetcher("animals", environment -> Arrays.asList(new Object(), new Object(), new Object())))
				.type("Query", builder -> builder.dataFetcher("plants", fetcher))
				.type("Animal", builder -> builder.dataFetcher("name", environment -> "cat"))
				.type("Plant", builder -> builder.dataFetcher("name", environment -> "petunia"))
				.build();

		SchemaGenerator schemaGenerator = new SchemaGenerator();
		GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(typeDefinitionRegistry, runtimeWiring);

		GraphQL build = GraphQL.newGraphQL(graphQLSchema)
				.queryExecutionStrategy(new AsyncSerialExecutionStrategy())
				.build();

		return build.execute("{animals{name} plants{name}}");
	}

}

```

Or add this test to AsyncSerialExecutionStrategyTest.groovy and notice the future never completes.

```groovy
    def "serial execution with errors"() {
        given:
        GraphQLSchema schema = schema(
                {env -> "hello"},
                {env -> "world"},
                {env -> throw new LinkageError("This should be handled")}
        )

        String query = "{hello, hello2, hello3}"
        def document = new Parser().parseDocument(query)
        def operation = document.definitions[0] as OperationDefinition

        def typeInfo = ExecutionStepInfo.newExecutionStepInfo()
                .type(schema.getQueryType())
                .build()

        ExecutionContext executionContext = new ExecutionContextBuilder()
                .graphQLSchema(schema)
                .executionId(ExecutionId.generate())
                .operationDefinition(operation)
                .instrumentation(SimplePerformantInstrumentation.INSTANCE)
                .valueUnboxer(ValueUnboxer.DEFAULT)
                .locale(Locale.getDefault())
                .graphQLContext(GraphQLContext.getDefault())
                .build()
        ExecutionStrategyParameters executionStrategyParameters = ExecutionStrategyParameters
                .newParameters()
                .executionStepInfo(typeInfo)
                .fields(mergedSelectionSet(['hello': mergedField(new Field('hello')), 'hello2': mergedField(new Field('hello2')), 'hello3': mergedField(new Field('hello3'))]))
                .build()

        AsyncSerialExecutionStrategy strategy = new AsyncSerialExecutionStrategy()
        when:
        def result = strategy.execute(executionContext, executionStrategyParameters)
        result.isDone()
        result.get()


        then:
        final ExecutionException exception = thrown()
        exception.message == 'java.lang.LinkageError: This should be handled'
    }
```

**Comment by andimarek** [comment]
catching a throwable is a bit complicated thing: it normally indicates that something is seriously wrong and often you can't really recover from it. It is not a normal exception?

Did you encounter this in real life? Which exception did you encounter?

**Comment by rlhagen** [comment]
Yes, I understand that.  But if the discussion is around catching Throwable... Currently, the code just crashes and it swallows the Error which is a way bigger issue than catching a Throwable. 

We have experienced this in real life mostly with `NoClassDefFoundError ` due an incorrect dependency loaded at runtime.   In this case catching the Throwable does the right thing since calling `join` or `get` on the future will throw a CompletionException which is far better than just swallowing the Error and causing the Future to be stuck indefinitely.

**Comment by rlhagen** [comment]
@andimarek Looks like this issue was fixed with @bbakerman commit for implementing [polymorphic](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/Async.java#L317) behavior.  Are you open to adding a test for the issue in the defect?

**Comment by bbakerman** [comment]
> Are you open to adding a test for the issue?


Very much so - thanks

**Comment by dondonz** [comment]
Hello if you'd like to add the test feel free to open a PR

As this issue was fixed with the 22.0 release I am closing this issue
